File: common/src/main/java/org/valkyrienskies/clockwork/content/contraptions/phys/bearing/PhysBearingData.java
package org.valkyrienskies.clockwork.content.contraptions.phys.bearing;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonIgnore;
import org.joml.Vector3d;
import org.joml.Vector3dc;
import org.valkyrienskies.core.apigame.constraints.*;

@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
public class PhysBearingData {

    public final Vector3dc bearingPosition;
    public final Vector3dc bearingAxis;
    public double bearingAngle;
    public float bearingRPM;
    public boolean locked;
    public long shiptraptionID;
    public boolean aligning;

    public VSAttachmentConstraint attachConstraint;
    @JsonIgnore
    public Integer attachID;

    public VSHingeOrientationConstraint hingeConstraint;
    public VSFixedOrientationConstraint angleConstraint;
    @JsonIgnore
    public Integer hingeID;

    // Default constructor for Jackson, should never be invoked manually
    @Deprecated
    public PhysBearingData() {
        this.bearingPosition = null;
        this.bearingAxis = null;
        this.shiptraptionID = -1;
    }

    public PhysBearingData(Vector3dc bearingPosition, Vector3dc bearingAxis, double bearingAngle, float bearingRPM, boolean locked, long shiptraptionID, VSConstraintAndId constraintAndId, VSConstraintAndId hingeConstraintAndId, VSConstraintAndId posDampConstraintAndId, VSConstraintAndId rotDampConstraintAndId) {
        this.bearingPosition = bearingPosition;
        this.bearingAxis = bearingAxis;
        this.bearingAngle = bearingAngle;
        this.bearingRPM = bearingRPM;
        this.locked = locked;
        this.shiptraptionID = shiptraptionID;
        this.attachConstraint = (VSAttachmentConstraint) constraintAndId.getVsConstraint();
        this.attachID = constraintAndId.getConstraintId();
        this.hingeConstraint = (VSHingeOrientationConstraint) hingeConstraintAndId.getVsConstraint();
        this.hingeID = hingeConstraintAndId.getConstraintId();
    }

    public void setAligning(boolean yn) {
        this.aligning = yn;
    }
}

File: common/src/main/kotlin/org/valkyrienskies/clockwork/content/contraptions/phys/bearing/data/PhysBearingData.kt
package org.valkyrienskies.clockwork.content.contraptions.phys.bearing.data

import com.fasterxml.jackson.annotation.JsonAutoDetect
import com.fasterxml.jackson.annotation.JsonIgnore
import org.joml.Vector3dc
import org.valkyrienskies.core.apigame.joints.VSFixedJoint
import org.valkyrienskies.core.apigame.joints.VSJointAndId
import org.valkyrienskies.core.apigame.joints.VSRevoluteJoint

@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
class PhysBearingData {
    //pos of bearing in subship coordinates
    val bearingPosition: Vector3dc?
    val bearingAxis: Vector3dc?
    var bearingAngle = 0.0
    var bearingRPM = 0f
    var locked = false
    var shiptraptionID: Long
    var aligning = false
    var attachConstraint: VSRevoluteJoint? = null

    @JsonIgnore
    var actualAngle = 0.0

    @JsonIgnore
    var attachID: Int? = null

    @JsonIgnore
    var hingeID: Int? = null

    @JsonIgnore
    var secondAttachId: Int? = null

    // Default constructor for Jackson, should never be invoked manually
    @Deprecated("")
    constructor() {
        bearingPosition = null
        bearingAxis = null
        shiptraptionID = -1
    }

    constructor(
        bearingPosition: Vector3dc?,
        bearingAxis: Vector3dc?,
        bearingAngle: Double,
        bearingRPM: Float,
        locked: Boolean,
        shiptraptionID: Long,
        constraintAndId: VSJointAndId,
    ) {
        this.bearingPosition = bearingPosition
        this.bearingAxis = bearingAxis
        this.bearingAngle = bearingAngle
        this.bearingRPM = bearingRPM
        this.locked = locked
        this.shiptraptionID = shiptraptionID
        attachConstraint = constraintAndId.joint as VSRevoluteJoint
        attachID = constraintAndId.jointId
    }
}